ckan/logic/action/file.py

```diff
@@ -0,0 +1,775 @@
+
+
+@logic.validate(schema.file_create)
+def file_create(context: Context, data_dict: dict[str, Any]) -> ActionResult.FileCreate:
+    """Create a new file.
+
+    The action passes uploaded file to the storage without strict
+    validation. File is converted into standard upload object and everything
+    else is controlled by storage. The same file may be accepted by one storage
+    and rejected by another, depending on its configuration.
+
+    The action is way too powerful to use it directly. The recommended approach
+    is to register a different action for handling specific type of uploads and
+    call the current action internally:
+
+    .. code-block:: python
+
+        def avatar_upload(context, data_dict):
+            logic.check_access("avatar_upload", context, data_dict)
+            storage = "avatars"
+            name = context["user"] + ".jpeg"
+            upload = data_dict["upload"]
+            return tk.get_action("file_create")(
+                Context(context, ignore_auth=True),
+                {"name": name, "storage": storage, "upload": upload},
+            )
+
+    When uploading a real file(or using ``werkqeug.datastructures.FileStorage``),
+    name parameter can be omited. In this case, the name of uploaded file is
+    used:
+
+    .. code-block:: sh
+
+        $ ckanapi action file_create upload@path/to/file.txt
+
+    When uploading a raw content of the file using bytes object, name is
+    mandatory:
+
+    .. code-block:: sh
+
+        $ ckanapi action file_create upload@<(echo -n "hello world") name=file.txt
+
+    .. note:: Requires storage with `CREATE` capability.
+
+    :param name: human-readable name of the file.
+        Defaults to filename of upload
+    :type name: str, optional
+    :param storage: name of the storage that will handle the upload.
+        Defaults to the configured ``default`` storage.
+    :type storage: str, optional
+    :param upload: content of the file as bytes, file descriptor or uploaded file
+    :type upload: bytes | file |
+        :py:class:`~werkqeug.datastructures.FileStorage` |
+        :py:class:`~ckan.lib.files.Upload`
+
+    :returns: file details.
+
+    """
+    logic.check_access("file_create", context, data_dict)
+    extras = data_dict.get("__extras", {})
+
+    try:
+        storage = files.get_storage(data_dict["storage"])
+    except files.exc.UnknownStorageError as err:
+        raise logic.ValidationError({"storage": [str(err)]}) from err
+
+    if not storage.supports(files.Capability.CREATE):
+        raise logic.ValidationError({"storage": ["Operation is not supported"]})
+
+    if "name" not in data_dict:
+        filename = data_dict["upload"].filename
+        if not filename:
+            msg = "Name is missing and cannot be deduced from upload"
+            raise logic.ValidationError({"upload": [msg]})
+        data_dict["name"] = filename
+
+    filename = secure_filename(data_dict["name"])
```

Comment:
looking at the werkzeug docs this might not be an ideal function to use. It's limited to ascii-only, removes paths and may result in an empty filename.